linera-core/src/client.rs

```diff
@@ -1370,18 +1346,250 @@ where
+                                continue;
+                            }
+                        }
+                        tracing::warn!("Skipping certificate {}: {}", hash, original_err);
+                    }
+                    break;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn find_missing_blobs_in_validator(
+        &self,
+        blob_ids: Vec<BlobId>,
+        chain_id: ChainId,
+        name: &ValidatorName,
+        node: &impl LocalValidatorNode,
+    ) -> Result<Vec<HashedBlob>, NodeError> {
+        let query = ChainInfoQuery::new(chain_id).with_manager_values();
+        let info = match node.handle_chain_info_query(query).await {
+            Ok(response) if response.check(name).is_ok() => Some(response.info),
+            Ok(_) => {
+                tracing::warn!("Got invalid response from validator");
+                return Ok(Vec::new());
+            }
+            Err(err) => {
+                tracing::warn!("Got error from validator: {}", err);
+                return Ok(Vec::new());
+            }
+        };
+
+        let mut missing_blobs = blob_ids;
+        let mut found_blobs = if let Some(info) = info {
+            let new_found_blobs = missing_blobs
+                .iter()
+                .filter_map(|blob_id| info.manager.pending_blobs.get(blob_id))
+                .map(|blob| (blob.id, blob.clone()))
+                .collect::<HashMap<_, _>>();
+            missing_blobs.retain(|blob_id| !new_found_blobs.contains_key(blob_id));
+            new_found_blobs.into_values().collect()
+        } else {
+            Vec::new()
+        };
+
+        found_blobs.extend(
+            LocalNodeClient::<S>::try_download_blobs_from(&missing_blobs, name, node).await,
+        );
+
+        Ok(found_blobs)
+    }
+
+    async fn update_local_node_with_blob_from(
+        &self,
+        blob_id: BlobId,
+        node: &impl LocalValidatorNode,
+    ) -> Result<(), ChainClientError> {
+        let Ok(last_used_by_hash) = node.blob_last_used_by(blob_id).await else {
+            return Err(ChainClientError::BlobNotFound(blob_id));
+        };
+
+        let certificate = node.download_certificate(last_used_by_hash).await?;
+
+        // This will download all ancestors of the certificate and process all of them locally.
+        self.receive_certificate(certificate).await?;
+        Ok(())
+    }
+
+    async fn receive_certificate_for_blob(&self, blob_id: BlobId) -> Result<(), ChainClientError> {
+        let validators = self.validator_nodes().await?;
+        let mut tasks = FuturesUnordered::new();
+        for (_, node) in validators.into_iter() {
+            let node = node.clone();
+            tasks.push(async move {
+                let last_used_hash = node.blob_last_used_by(blob_id).await.ok()?;
+                let cert = node.download_certificate(last_used_hash).await.ok()?;
```

Comment:
@andresilva91: Sorry, I missed this in the review. We still have to check here that the certificate actually does read or publish the expected blob! Otherwise a faulty node can send us _any_ certificate and we'll return from this function and think we've succeeded.

Reply:
Good catch! PR #2321 should fix it